common/wuanalysis/anawu.cpp

```diff
@@ -0,0 +1,429 @@
+    {
+        StringBuffer name;
+        name.append('@').append(queryWuAttributeName(attr));
+        scope->setProp(name, value);
+    }
+    virtual void noteHint(const char * kind, const char * value)
+    {
+        throwUnexpected();
+    }
+    IPropertyTree * scope;
+};
+
+
+//-----------------------------------------------------------------------------------------------------------
+WorkunitAnalyser::WorkunitAnalyser(WuAnalyseOptions & _options) : root("", nullptr), options(_options)
+{
+    gatherRules(rules);
+}
+
+void WorkunitAnalyser::check(const char * scope, IWuActivity & activity)
+{
+    if (activity.queryThorActivityKind()!=TAKhashdistribute)
+    {
+        if (activity.getStatRaw(StTimeLocalExecute, StMaxX) < options.minInterestingTime)
+            return;
+    }
+
+    Owned<PerformanceIssue> issue;
+    ForEachItemIn(i, rules)
+    {
+        if (rules.item(i).isCandidate(activity))
+        {
+            if (!issue)
+                issue.setown(new PerformanceIssue);
+            if (rules.item(i).check(*issue, activity, options))
+            {
+                if (issue->cost >= options.minCost)
+                {
+                    issue->scope.set(scope);
+                    issues.append(*issue.getClear());
+                    break;
+                }
+                else
+                {
+                    issue.clear();
+                }
+            }
+        }
+    }
+}
+
+void WorkunitAnalyser::check(IConstWorkUnit * wu)
+{
+    WuScopeFilter filter;
+    filter.addOutputProperties(PTstatistics).addOutputProperties(PTattributes);
+    filter.finishedFilter();
+    collateWorkunitStats(wu, filter);
+    root.connectActivities();
+    // root.trace();
+    root.applyRules(*this);
+    limitResultsPerScope();
+    printWarnings();
+}
+
+void WorkunitAnalyser::collateWorkunitStats(IConstWorkUnit * workunit, const WuScopeFilter & filter)
+{
+    Owned<IConstWUScopeIterator> iter = &workunit->getScopeIterator(filter);
+    ForEach(*iter)
+    {
+        try
+        {
+            WuScope * scope = selectFullScope(iter->queryScope());
+
+            StatsGatherer callback(scope->queryAttrs());
+            scope->queryAttrs()->setPropInt("@stype", iter->getScopeType());
+            iter->playProperties(callback);
+        }
+        catch (IException * e)
+        {
+            e->Release();
+        }
+    }
+}
+
+// Only keep highest cost result per scope
+void WorkunitAnalyser::limitResultsPerScope()
+{
+    issues.sort(compareIssuesScopeOrder);
+
+    int prevIssue = 0;
+    int ordinality = issues.ordinality();
+    int indx = 1;
+    while (indx < ordinality)
+    {
+        if (issues.item(indx).compareScope(issues.item(prevIssue))==0)   // Same scope so remove one
+        {
+            if (issues.item(indx).compareCost(issues.item(prevIssue))<0) // remove the lower cost of the 2 issues
+            {
+                prevIssue = indx;
```

Comment:
I think this is wrong - since indx-1 is removed - so the current indx item will now be at indx-1.  prevIssue is always indx-1

Reply:
Function removed. No longer relevant.